#### Import Libraries and Dependencies

In [1]:
import pandas as pd
from pathlib import Path

#### Read in Files 

In [2]:
# Import data
tsla_path = Path('../Resources/tsla_stock_volume.csv')

# Read in data and index by date
tsla_data = pd.read_csv(
    tsla_path, 
    index_col='Date',     
    parse_dates=True, 
    infer_datetime_format=True
)
tsla_data

close  volume
Date                               
2010-06-29 10:00:00    3.62  412770
2010-06-29 11:00:00    3.67   98220
2010-06-29 12:00:00    3.82   84465
2010-06-29 13:00:00    4.40  159355
2010-06-29 14:00:00    4.79  161970
...                     ...     ...
2021-02-08 10:00:00  864.26   10684
2021-02-08 11:00:00  865.49    9550
2021-02-08 12:00:00  858.38   10333
2021-02-08 13:00:00  858.72   12833
2021-02-08 14:00:00  863.30   15212

[19264 rows x 2 columns]

#### Slice DateTimeIndex Data

In [3]:
# Slice the Data to One Specific Month
volume_jan_2021 = tsla_data.loc['2021-01-01':'2021-01-31']
volume_jan_2021

close  volume
Date                               
2021-01-04 08:00:00  738.55   70474
2021-01-04 09:00:00  731.20   61620
2021-01-04 10:00:00  730.98   40010
2021-01-04 11:00:00  724.06   45447
2021-01-04 12:00:00  733.13   32008
...                     ...     ...
2021-01-29 10:00:00  819.00   15860
2021-01-29 11:00:00  790.72   48941
2021-01-29 12:00:00  790.88   27588
2021-01-29 13:00:00  789.09   20261
2021-01-29 14:00:00  793.31   29397

[133 rows x 2 columns]

#### Analyze DateTimeIndex Data with Groupby

In [4]:
# Calculate the total number of shares traded for the month of January, 2021
jan_2021_volume = volume_jan_2021['volume'].sum()
jan_2021_volume

4161226

In [5]:
tesla_volume = tsla_data['volume']
tesla_volume.head()

Date
2010-06-29 10:00:00    412770
2010-06-29 11:00:00     98220
2010-06-29 12:00:00     84465
2010-06-29 13:00:00    159355
2010-06-29 14:00:00    161970
Name: volume, dtype: int64

In [6]:
# Specify the way you want to group things -- here we are using the datetimeindex
groupby_levels = [tesla_volume.index.year, tesla_volume.index.month]

# Then Groupby that, choosing an aggregation function
total_monthly_volume = tesla_volume.groupby(by=groupby_levels).sum()
total_monthly_volume

Date  Date
2010  6       1768530
      7       3162325
      8        722650
      9        870375
      10       316900
               ...   
2020  10      5535663
      11      4984508
      12      6890093
2021  1       4161226
      2        671076
Name: volume, Length: 129, dtype: int64

In [7]:
# We can do summary statistics on the aggregated data we just created
# For Example: View the median amount of monthly shares traded
median_monthly_volume = total_monthly_volume.median()
median_monthly_volume

4677177.0

In [8]:
# Compare the shares traded in January 2021 to the median amount that get traded each month
jan_2021_volume / median_monthly_volume

0.8896875187746797

The result is that January 2021 volume was about 12% below that of the typical amount of Tesla shares that get traded each month. 